qutip/core/data/expm.py

```diff
@@ -12,12 +12,16 @@
 
 def expm_csr(matrix: CSR) -> CSR:
     if matrix.shape[0] != matrix.shape[1]:
-        raise TypeError("can only exponentiate square matrix")
+        raise ValueError("can only exponentiate square matrix")
     if isdiag_csr(matrix):
-        out = matrix.copy()
-        sci = out.as_scipy()
-        sci.data[:] = np.exp(sci.data)
-        return out
+        sci = matrix.as_scipy()
+        # Note that expm1 is an elemetwise operation that maps 0->0 which
+        # involves no operation at all. Since we know that the input is
+        # diagonal, elementwise expm1 in the whole matrix has the same
+        # computational cost as doing only a expm1 in the diagonal
+        # (elementwise).
+        sci = sci.expm1() + scipy.sparse.diags(np.ones(matrix.shape[0]))
+        return CSR(sci)
```

Comment:
This is a good catch, thanks.  We could consider optimising this algorithm a little - it involves several allocations, three instantiations of the slow `scipy.sparse.csr_matrix`, and one CSR addition, which aren't super ideal.  I would imagine something like:
```python
        matrix_sci = matrix.as_scipy()
        data = np.ones(matrix.shape[0], dtype=np.complex128)
        data[matrix_sci.indices] += np.expm1(matrix_sci.data)
        return _data.CSR(
            (
                data,
                np.arange(matrix.shape[0], dtype=_data.base.idxint_dtype),
                np.arange(matrix.shape[0] + 1, dtype=_data.base.idxint_dtype),
            ),
            shape=matrix.shape,
            copy=False,
        )
```
would be more efficient.  We could avoid a couple more allocations if we rewrote this in Cython as well (we wouldn't need to allocate in the `np.expm1` call and we could avoid a couple of internal `CSR.__init__` checks), but that's probably not really necessary.  I would expect the overhead of this method to be reasonably low anyway.

Reply:
I can go  with this implementation. Thanks for the input!

Reply:
Please do make sure it's right, and it's probably worth doing a proper benchmark! The one I wrote _looks_ right to me  and it makes rather fewer allocations (plus numpy instantiations are cheaper), so I expect it it's noticeably faster, but I'm not certain.